### PR TITLE
Improve translation loading

### PR DESCRIPTION
**Pull request by Holzhaus**
While using developer mode, I saw some suspicious log messages regarding failed translations when using the `en` locale. When looking at the code I realized that mixxx doesn't even attempt to load them (which makes sense) but the error message is misleading.

I cleaned up the translation handling a bit, added some error messages and made them more meaningful. I based this on 2.3 but If you want to postpone this for 2.4 we can do that as well.

### CODE REVIEW

File: src/util/translations.h [hunk 1]
```diff
@@ -13,92 +13,101 @@ namespace mixxx {
...
+                }
+                QLocale::setDefault(customLocale);
+            }
         }
 
-        if (userLocale.isEmpty()) {
-            QLocale::setDefault(QLocale(systemLocale));
-        } else {
-            QLocale::setDefault(QLocale(userLocale));
+        QLocale locale = QLocale();
```
Comment (daschuer) [inline src/util/translations.h:52]: This assignment is redundant.
Please add a comment that we investigate the default value.

**Review by daschuer — CHANGES_REQUESTED** [review]
2.3 is fine for me. I have left some comments.

File: src/util/translations.h [hunk 2]
```diff
@@ -13,92 +13,101 @@ namespace mixxx {
...
+        const QString translationsPath =
+                pConfig->getResourcePath() + QStringLiteral("translations/");
 
-        // Load Qt base translations
-        QString userLocale = forcedLocale;
-        QLocale systemLocale = QLocale::system();
+        // Set the default locale, if specified via command line or config
+        // file. If neither is the case, the default locale will be the
+        // system's locale.
+        {
```
[the remark this answers is not in the document]
Comment (Holzhaus) [reply src/util/translations.h:24]: It's not really necessary but `customLocaleString` will be dropped after it. I can just remove it if it's confusing.

File: src/util/translations.h (hunk 1 above)
Comment (Holzhaus) [reply src/util/translations.h:52]: > Please add a comment that we investigate the default value.

What do you mean?

From the [Qt docs](https://doc.qt.io/qt-5/qlocale.html#QLocale):
> Constructs a QLocale object initialized with the default locale. If no default locale was set using setDefault(), this locale will be the same as the one returned by system().

Comment (Holzhaus) [reply src/util/translations.h]: I added that as a comment.

File: src/util/translations.h [hunk 3]
```diff
@@ -13,92 +13,99 @@ namespace mixxx {
...
-        } else {
-            delete qtTranslator;
-        }
+        installTranslations(pApp,
+                locale,
+                QStringLiteral("qt"),
+                QLibraryInfo::location(QLibraryInfo::TranslationsPath));
 
         // Load Qt translations for this locale from the Mixxx translations
         // folder.
```
Comment (Holzhaus) [inline src/util/translations.h:78]: Is this even relevant anymore? I don't think there exist...

```
$ ls -l ../res/translations/ | grep qt | wc -l
0
```

Comment (Holzhaus) [reply src/util/translations.h:78]: Wait, is this only relevant on Windows?

Comment (Holzhaus) [reply src/util/translations.h:79]: For these translations, I changed to loglevel to Debug instead of warning.

**Comment by daschuer** [comment]
Warning [Main]: Unsupported locale "ceb"

When I use "System" I get confusing messages on Ubuntu Focal: 

```
Debug [Main]: Found and will use default keyboard preset "/home/daniel/workspace/2.3/res/keyboard/en_US.kbd.cfg"
Debug [Main]: Loading resources from  "/home/daniel/workspace/2.3/res/"
Debug [Main]: Loaded "qt" translations for locale "de_DE" from "/usr/share/qt5/translations"
Debug [Main]: Failed to load "qt" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
Debug [Main]: Loaded "mixxx" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
```

**Comment by Holzhaus** [comment]
> Warning [Main]: Unsupported locale "ceb"

I cannot reproduce this. I guess your config was messed up and contained an invalid string for the locale setting.

> When I use "System" I get confusing messages on Ubuntu Focal:
> 
> ```
> Debug [Main]: Found and will use default keyboard preset "/home/daniel/workspace/2.3/res/keyboard/en_US.kbd.cfg"
> Debug [Main]: Loading resources from  "/home/daniel/workspace/2.3/res/"
> Debug [Main]: Loaded "qt" translations for locale "de_DE" from "/usr/share/qt5/translations"
> Debug [Main]: Failed to load "qt" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
> Debug [Main]: Loaded "mixxx" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
> ```

What is confusing about it? This looks correct to me.

It's normal that loading `qt` translations from the mixxx translation directory fails, i think that is only used on windows (an macOS maybe?). See https://github.com/mixxxdj/mixxx/pull/3826/commits/e847f58f5a268b104206a271ab709ae5ff7981c1.

**Comment by esbrandt** [comment]
> Warning [Main]: Unsupported locale "ceb"

Delete 
https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.qm
https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.ts
Both files have never been touched since ages, and the language is not even supported on Transifex.

**Comment by Holzhaus** [comment]
> > Warning [Main]: Unsupported locale "ceb"
> 
> Delete
> https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.qm
> https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.ts
> Both files have never been touched since ages, and the language is not even supported on Transifex.

Done.

**Comment by Holzhaus** [comment]
Anything else to do here?

**Comment by esbrandt** [comment]
LGTM

**Comment by Holzhaus** [comment]
ping @daschuer

**Review by uklotzde — APPROVED** [review]
Very helpful comments. LGTM

@daschuer Ping

**Comment by Holzhaus** [comment]
> Why do we tray to load the qt translations from the res path when we already have loaded them from the system? Can we make the local lookup conditionally? Or load directly from the correct location target depending. 
> 
> On windows we install also qtbase and qthelp translations among others in the translations folder but the corresponding install statement is missing. I am afraid they are not loaded, are they? Are they loaded under Windows. 

I didn't change this. If it worked before, I see no reason why it should not work now.

We could check if the system install translations call succeeded and skip the other one, but I don't want to introduce a bug so shortly before release (is it possible that there incomplete system translations for example?). 

> Only the en-GB translation in the translation folder used an underscore instead of a dash like the others. Does this work as well?  
> In the /usr/share/qt5/translations folder the country postfix is always separated by _

I think so. Just try with one of these:

    $ LANG=de_DE ./mixxx
    $ ./mixxx --locale de_DE

I didn't notice any issues when testing this.

**Comment by uklotzde** [comment]
> Why do we tray to load the qt translations from the res path when we already have loaded them from the system? Can we make the local lookup conditionally? Or load directly from the correct location target depending.
> 
> On windows we install also qtbase and qthelp translations among others in the translations folder but the corresponding install statement is missing. I am afraid they are not loaded, are they? Are they loaded under Windows.
> 
> Only the en-GB translation in the translation folder used an underscore instead of a dash like the others. Does this work as well?
> In the /usr/share/qt5/translations folder the country postfix is always separated by _

https://doc.qt.io/qt-5/qlocale.html#QLocale-1

> The separator can be either underscore or a minus sign.

**Comment by daschuer** [comment]
I am afraid really have a regression here. I am sorry that this PR is such hard.  

I have selected "Spanish (Mexico)" from the interface preferences in Mixxx 2.3. After restart, the Auto DJ tree knot shows the correct "DJ Automatico" from the mixxx_es-MX.ts file. After merging this branch, the Auto DJ knot is renamed to "Auto DJ" which is part of the mixxx_es.ts

By using and intermediate QLocale, Qt turns the - to _ and the translation is no longer found.

**Comment by Holzhaus** [comment]
You're right, but I don't think this is a regression. Before, it also didn't work properly because it was required to use the correct separator (underscore vs. dash) to get the correct locale. That's why I didn't notice. And for System language, this also didn't work I guess.

There is no way to use the dash as delimiter without building the file name yourself:
https://doc.qt.io/qt-5/qtranslator.html#load-1

Instead, we should use the standard underscore for our translations IMHO. If you agree I'll rename the translation files, but we also need to figure out to configure tx correctly.

**Comment by daschuer** [comment]
Renaming the files, works for me as a 2.3.0 band aid.
The loading of the additional qt translation can be postponed.

**Comment by Holzhaus** [comment]
@daschuer Please check again, should be fixed now.

File: .tx/config [hunk 4]
```diff
@@ -1,6 +1,5 @@
...
...hu-HU, hy_AM: hy-AM, id_ID: id-ID, ig_NG: ig-NG, is_IS: is-IS, it_CH: it-CH, it_IT: it-IT, ja_JP: ja-JP, ka_GE: ka-GE, kk_KZ: kk-KZ, kl: kl-GL, km_KH: km-KH, kn_IN: kn-IN, ko_KR: ko-KR, ku_IQ: ckb-IQ, ky: ky-KG, lb: lb-LU, lo_LA: lo-LA, lt_LT: lt-LT, lv_LV: lv-LV, mi: mi-NZ, mk_MK: mk-MK, ml_IN: ml-IN, mn_MN: mn-MN, mr_IN: mr-IN, ms_MY: ms-MY, mt_MT: mt-MT, my_MM : my-MM, nb_NO: nb-NO, ne_NP: ne-NP, nl_BE: nl-BE, nl_NL: nl-NL, nn_NO: nn-NO, nso: nso-ZA, oc: oc-FR, or_IN: or-IN, pa_IN: pa-IN, pl_PL: pl-PL, ps: ps-AF, pt_BR: pt-BR, pt_PT: pt-PT, ro_RO: ro-RO, ru_RU: ru-RU, rw: rw-RW, sah: sah-RU, sd: sd-PK, se: se-FI, se_NO: se-NO, se_SE: se-SE, si_LK: si-LK, sk_SK: sk-SK, sl_SI: sl-SI, sma: sma-SE, sq_AL: sq-AL, sr@latin: sr-YU, sr_BA: sr-BA, sr_CS: sr-CS, sr_ME: sr-ME, sr_RS: sr-RS, sv_FI: sv-FI, sv_SE: sv-SE, sw_KE: sw-KE, ta_IN: ta-IN, ta_LK: ta-LK, te_IN: te-IN, tet: tet-TL, tg_TJ: tg-TJ, th_TH: th-TH, tk_TM: tk-TM, tl_PH: tl-PH, tn: tn-ZA, tr_TR: tr-TR, tt: tt-RU, tzm: tzm-DZ, ug: ug-CN, uk_UA: uk-UA, ur_PK: ur-PK, uz@Latn: uz-UZ, uz@Cyrl: uzb-UZ, vi_VN: vi-VN, wo_SN: wo-SN, xh: xh-ZA, yo: yo-NG, zh_CN: zh-CN, zh_HK: zh-HK, zh_MO: zh-MO, zh_SG: zh-SG, zh_TW: zh-TW, zu_ZA: zu-ZA
```
Comment (daschuer) [inline .tx/config:3]: Thank you for digging down to the root cause.

**Comment by daschuer** [comment]
I can confirm it is working now.  
As a side effect the selection falls back to system, because the dash version in the preferences is no longer found. 
But IMHO we can effort this, because the old state was kind of broken anyway. 

Is this issue worth a changelog entry?

**Review by daschuer — APPROVED** [review]

**Pull request merged**